### PR TITLE
SLING-10011 : Use JackrabbitSession.getParentOrNull when resolving parent JCR node in JcrResourceProvider#getParent

### PR DESCRIPTION
**Pull request by anchela**

### CODE REVIEW

**Comment by anchela** [comment]
@cziegeler , @joerghoh , picking up SLING-10011 since the oak improvement is now officially released.

we might need to add some extra checks if _JackrabbitSession.getParentOrNull_ is available though... i didn't add that in a first step. but i am aware that this kind of checks is done in other parts of Sling. wdyt?

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrItemResourceFactory.java
Comment (joerghoh) [inline src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrItemResourceFactory.java:215]: could this also be because of access control? Would it make sense to mention it as well?

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrResourceProvider.java
Comment (joerghoh) [inline src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrResourceProvider.java:366]: If we knew here that it is already a Jackrabbit-Repo, we would not need to calculate the parentPath for this codepath. This would be only necessary when line 371 is reached.

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrItemResourceFactory.java
Comment (anchela) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrItemResourceFactory.java:215]: most likely it is due to lack of permission.... it's the original message and i didn't alter that one. i would leave out any confusing attempt to explain what could potentially be wrong.... but that's my take.

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrResourceProvider.java
Comment (anchela) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrResourceProvider.java:366]: not quite.... but i fell into the same trap..... but looking at ResourceUtil.getParent it does some extra magic like returning null if the item is the root node.
also the calculated parent path is used for the fallback when no JackrabbitSession is available.
so, i decided to leave the code instead of having it duplicated with all the logic involved.... but maybe there is a way to simplify it.... but ResourceUtil.getParent looks quite complex and i didn't want to introduce nasty side effects by replacing it by PathUtil.getParent or some other utility i am familiar with.

File: src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrItemResourceFactory.java
Comment (anchela) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrItemResourceFactory.java:215]: see also line 198

Comment (anchela) [reply src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrItemResourceFactory.java]: hopefully better now :)

**Review by joerghoh — APPROVED** [review]

**Comment by joerghoh** [comment]
The only problem I could see is the updated dependency to Oak 1.42 (which can cause problems on downstream applications, which still use an older version, e.g. AEM 6.5).

**Comment by reschke** [comment]
Well, we could backport to Oak 1.22.*.

**Comment by anchela** [comment]
@joerghoh , yes. that's what i meant with my initial comment, when i said we might need some extra check. i remember having seen something like that in sling code, but i don't recall where that was.

@reschke , the problem with back-porting API extensions is usually that we need to be extra careful with the semantic versioning.... that's why i usually try to avoid it.

**Comment by reschke** [comment]
> @reschke , the problem with back-porting API extensions is usually that we need to be extra careful with the semantic versioning.... that's why i usually try to avoid it.

Understood, but in this case I believe it's possible; at least for 1.22.

**Comment by cziegeler** [comment]
I think we shouldn't worry too much about updating older versions, that can be handled in various ways. Of course a 1.2.2 would make that story much easier

**Comment by anchela** [comment]
@reschke would you have time to backport the latest improvements of jackrabbit api to the 1.22. branch (there were a few other things as well in the security area like the new `PrivilegeCollection` that is super handy and much faster when it comes to evaluating privileges without manually resolving the aggregation (like it is e.g. done in https://github.com/apache/sling-org-apache-sling-jcr-repoinit/blob/master/src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java#L593).

**Comment by reschke** [comment]
Yes, I can do that. Do we want a separate OAK ticket, or do we treat this as regular backport?

**Comment by anchela** [comment]
@cziegeler not sure i got your comment..... does it mean you are ok with the proposed fix and not being able to update _sling-org-apache-sling-jcr-resource_ older sling instances to the latest version unless they also update to the latest oak version or we make sure oak 1.22 also exposes the new API?

if that is the case i would check again why the quality gate is failing and eventually merge the fix.
wdyt?

**Comment by cziegeler** [comment]
@anchela  I'm ok with requiring older installations to update to latest Oak in order to get this improvement.
If Oak wants to do add this to the 1.22 api, then this is great, but I don't see it as a strict requirement

**Comment by anchela** [comment]
@reschke , i am ok with regular back port.... if you prefer a ticket i can create one. whatever is easier for you.

**Comment by anchela** [comment]
@cziegeler , thanks. then i am only waiting for your approval.... in the mean time i added a test for JcrResourceProvider.getParent. there was....  none..... 🙀 

and while writing the test i was wondering why the `getParent` method doesn't take advantage of the fact that  `JcrResourceProvider.getResource` already takes a parent-resource..... but well.... i am not going to dig into that in the scope of this improvement. but might be worth a second thought.

**Review by cziegeler — APPROVED** [review]
lgtm, thanks

**Pull request merged**
